R/convergence.R

```diff
@@ -316,6 +316,144 @@ quantile2 <- function(x, probs = c(0.05, 0.95), names = TRUE, ...) {
+#' the random seed is fixed, it will generally produce a different result. To minimise
+#' the implications of this stochasticity, it is recommended to repeatedly run this
+#' function to calculate a distribution of R*; alternatively, an approximation to this
+#' distribution can be obtained by setting uncertainty = T.
+#'
+#' By default, a random forest classifier is used, which tends to perform best for
+#' target distribution dimensionalities of around 4 and above. For lower dimensional
+#' targets, gradient boosted models (called via `method=\"gbm\"`) tend to have a
+#' higher classification accuracy. On a given MCMC sample, it is recommended to
+#' try both of these classifiers.
+#'
+#' @return A single numeric R* value (by default) or a vector of values (if
+#' uncertainty is true).
+#'
+#' @references Ben Lambert, Aki Vehtari (2020) R*: A robust MCMC convergence
+#' diagnostic with uncertainty using gradient-boosted machines
+#' \emph{arXiv preprint} \code{arXiv:TBD}
+#'
+#' @examples
+#' x <- example_draws("eight_schools")
+#' rstar(x)
+#' rstar(x, split_chains = F)
+#' rstar(x, method = "gbm")
+#' # can pass additional arguments to methods
+#' rstar(x, method = "gbm", verbose = F)
+#'
+#' # with uncertainty, returns a vector of R* values
+#' hist(rstar(x, uncertainty = T))
+#' hist(rstar(x, uncertainty = T, nsim = 100))
+#'
+#' # can use other classification methods in caret library
+#' rstar(x, method = "knn")
+#'
+#' @export
+rstar <- function(x, split_chains=T, uncertainty=F, method=NULL, hyperparameters=NULL,
+                            training_proportion=0.7, nsim=1000, ...){
+
+  if(!requireNamespace("caret", quietly = T))
+    stop("Package \"caret\" needed for this function to work. Please install it.")
+  require(caret)
+
+  nsim <- round(nsim)
+  if(nsim < 1)
+    stop("nsim must exceed 1.")
+
+  if(training_proportion <= 0 || training_proportion >= 1)
+    stop("training_proportion must be greater than zero and less than 1.")
+
+  x <- as_draws_df(x)
+  if(split_chains)
+    x <- split_chains_draws_df(x)
+
+  # if only 1 param, add in a column of random noise
+  nparams <- nvariables(x)
+  if(nparams==1)
+    x$V_new <- rnorm(nrow(x))
+
+  # create training / testing sets
+  x$.chain <- as.factor(x$.chain)
+  rand_samples <- sample(1:nrow(x), training_proportion * nrow(x))
+  training_data <- x[rand_samples, ]
+  testing_data <- x[-rand_samples, ]
+
+  # choose hyperparameters
+  if(is.null(method)) {
+    method <- "rf"
+    caret_grid <- data.frame(mtry=floor(sqrt(nparams)))
+  } else if(is.null(hyperparameters) && method=="gbm") {
+    caret_grid <- data.frame(interaction.depth=3,
+                         n.trees = 50,
+                         shrinkage=0.1,
+                         n.minobsinnode=10)
+  } else {
+    caret_grid <- hyperparameters
+  }
+
+  # remove iteration / draws columns and fit classifier
+  training_data <- training_data[, c(variables(training_data), ".chain")]
```

Comment:
As removing .iteration and .draw invalidates a draws_df object, I recommend removing the draws classes from the object to avoid any unexpected behavior. This works, for example. via `class(training_data) <- class(training_data)[-(1:2)]

Reply:
Done.